leakpro/attacks/utils/boosting.py

```diff
@@ -0,0 +1,212 @@
+                                    0, 1).squeeze()
+
+        logits = self.softmax_logits(logits)
+        target_logits = self.softmax_logits(target_logits)
+
+        if self.online:
+            for i, (logit, mask, label) in tqdm(enumerate(zip(logits, self.in_indices_mask, self.audit_data_labels)),
+                                                total=len(logits),
+                                                desc="Calculating memorization score",
+                                                leave=False
+                                                ):
+                self.memorization_score[i] = np.mean(logit[mask, label]) - np.mean(logit[~mask, label])
+        else:
+            for i, (logit, target_logit, label) in tqdm(enumerate(zip(logits, target_logits, self.audit_data_labels)),
+                                                total=len(logits),
+                                                desc="Calculating memorization score",
+                                                leave=False
+                                                ):
+                self.memorization_score[i] = np.mean(target_logit[label]) - np.mean(logit[:, label])
+
+    def _privacy_score(self:Self) -> None:
+        """Run privacy score enhancement.
+
+        Privacy score enhances the attack performance by only including vulnerable data points.
+        The privacy score definition can be found in "Membership Inference Attacks From First Principles" by Carlini et al.
+        """
+
+        logits_function = ModelRescaledLogits()
+        logits = np.swapaxes(logits_function(self.shadow_models, self.handler, self.audit_data_indices, self.batch_size), 0, 1)
+        target_logits = np.swapaxes(logits_function([self.target_model], self.handler, self.audit_data_indices, self.batch_size),\
+                                    0, 1).squeeze()
+
+        privacy_score = []
+
+        # From "Membership Inference Attacks From First Principles", fixed variance is used when number of shadow models < 64
+        if len(self.shadow_models) < 64:
+            in_std, out_std = np.std(logits[self.in_indices_mask].flatten()), np.std(logits[~self.in_indices_mask].flatten())
+
+        for (logit, target_logit, mask) in tqdm(zip(logits, target_logits, self.in_indices_mask),
+                                                total=len(logits),
+                                                desc="Calculating privacy score",
+                                                leave=False
+                                                ):
+            in_mean, out_mean = np.mean(logit[mask]), np.mean(logit[~mask])
+
+            if len(self.shadow_models) >= 64:
+                in_std, out_std = np.std(logit[mask]), np.std(logit[~mask])
+
+            if self.online:
+                privacy_score.append(np.abs(in_mean-out_mean)/(in_std+out_std+1e-30))
+            else:
+                privacy_score.append(np.abs(target_logit-out_mean)/(2*out_std+1e-30))
```

Comment:
And for the privacy score assumptions @johanos1 
\#Assumptions
\# 1. target_logit is a good aproximation for the missing IN-samples
\# 2. out_std ~= in_std => in_std+out_std ~= 2*out_std